benchmarks/distributed/rpc/rl/agent.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+
+import torch.distributed.rpc as rpc
+from torch.distributed.rpc import RRef, remote
+from torch.distributions import Categorical
+
+from Observer import ObserverBase
+
+OBSERVER_NAME = "observer{}"
+
+gamma = 0.99
+
+
+class Policy(nn.Module):
+    def __init__(self, n_layers=0, in_features=10, out_features=10):
+        super(Policy, self).__init__()
+        self.n_layers = n_layers
+        self.in_features = in_features
+        self.out_features = out_features
+
+        self.model = nn.Sequential(
+            *[nn.Linear(in_features, out_features) for _ in range(n_layers + 1)])
+        self.saved_log_probs = []
+        self.rewards = []
+
+    def forward(self, x):
+        action_scores = self.model(x)
+        return F.softmax(action_scores, dim=1)
+
+
+class AgentBase:
+    def __init__(self):
+        self.id = rpc.get_worker_info().id
+        self.running_reward = 0
+        self.policy = Policy()
+        self.optimizer = optim.Adam(self.policy.parameters(), lr=1e-2)
+        self.eps = np.finfo(np.float32).eps.item()
+
+        self.ob_rrefs = []   # Observer RRef
+        self.rewards = {}
+        self.saved_log_probs = {}
+
+    def set_world(self, world_size):
+        self.world_size = world_size
+
+        for rank in range(2, world_size):
+            ob_info = rpc.get_worker_info(OBSERVER_NAME.format(rank))
+            self.ob_rrefs.append(remote(ob_info, ObserverBase))
+            self.rewards[ob_info.id] = []
+            self.saved_log_probs[ob_info.id] = []
+
+    def select_action(self, observer_id, state, reward):
+        self.rewards[observer_id].append(reward)
+
+        state = torch.from_numpy(state).float().unsqueeze(0)
+        probs = self.policy(state)
+        m = Categorical(probs)
```

Comment:
It is up to you whether you would like to keep this `Categorical` or simply store the outputs from the policy. Since we are mainly focusing on testing RPC efficiency, you do not have to faithfully mimic what a RL agent would do. But it also does not hurt to have this. So, please feel free to make your own choice here.